lib/SILOptimizer/Mandatory/TFUtilities.cpp

```diff
@@ -1510,6 +1519,126 @@ SILInstruction *SILTensorOpInfo::canonicalizeOperands(
   return newInst;
 }
 
+//===----------------------------------------------------------------------===//
+// GraphOperationDecoder implementation
+//===----------------------------------------------------------------------===//
+
+/// Return the device attribute associated with `inst`, which is required to
+/// exist.
+StringRef GraphOperationDecoder::getDeviceString() const {
+  auto attr = inst->getAttribute(DEVICE_ATTR);
+  assert(attr.hasValue() && "Tensor op instruction has no device string");
+  return attr.getValue().getStringValue();
+}
+
+/// Decode the name of a graph_op into its TensorFlow op name and a list of
+/// information about the operands.
+StringRef GraphOperationDecoder::
+decodeName(SmallVectorImpl<InputMarker> &inputInfo) {
+  auto name = inst->getName().str();
+  auto pos = name.find(',');
+  auto opName = name.substr(0, pos);
+
+  // TODO: Fill in inputInfo when we support input lists.
+
+  return opName;
+}
+
+/// Given an attribute name like foo$dtype, decode the name and the class.
+std::pair<StringRef, SILTensorOpInfo::OperandClass>
+GraphOperationDecoder::decodeAttributeName(Identifier name) {
+  auto nameStr = name.str();
+  // Figure out what the suffix is (if any).
+  auto dollarLoc = nameStr.find('$');
+
+  auto opClass = OperandClass::Normal;
+  if (dollarLoc != StringRef::npos) {
+    auto suffix = nameStr.drop_front(dollarLoc+1);
+    opClass = SILTensorOpInfo::getOperandClass(suffix).getValue();
+  }
+
+  // Slice the suffix off the attribute name and add the decoded version.
+  return { nameStr.substr(0, dollarLoc), opClass };
+}
+
+
+//===----------------------------------------------------------------------===//
+// Device Partitioning Utilities
+//===----------------------------------------------------------------------===//
+
+/// Scan the specified function, looking for logic that configures the current
+/// graph.
+GraphGlobalConfiguration
+GraphGlobalConfiguration::getForFunction(SILFunction &fn) {
```

Comment:
it's great that we are extending the config from global to per-function. Some suggestions/questions:

1. should we rename GraphGlobalConfiguration to sth like DeviceConfiguration or DevicePlacementConfiguration? The name "global" is no longer appropriate.

2. should we extend the unit testing on TensorFlow.enableGPU() / enableTPU() for the per-function case? e.g. if foo() calls bar(), and we see such a statement in both foo() and bar(), i guess the one in bar() should error out, when we are partitioning foo(), and bar() is inlined into foo() by our deabstraction? 

It'd be fine to leave a TODO for #2 in this PR.